From 2e679c66554cc0205779a72b82ddcc0175f571a6 Mon Sep 17 00:00:00 2001
From: Gregory Petrosyan <gregory.petrosyan@gmail.com>
Date: Sun, 23 May 2010 20:02:58 +0400
Subject: [PATCH] Fix format_print()'s buffer width calculation

This solves reported memory corruptions and crashes.
Looks like this bug was uncovered due to a GCC upgrade.

Reported-by: Vlastimil Kotas <vlastikx@kralikbobek.cz>
Reported-by: evinupavi662@arcor.de
Triaged-by: Xavier Chantry <chantry.xavier@gmail.com>
Acked-by: Xavier Chantry <chantry.xavier@gmail.com>
Signed-off-by: Gregory Petrosyan <gregory.petrosyan@gmail.com>
---
 ui_curses.c |    8 +++++---
 1 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/ui_curses.c b/ui_curses.c
index 0a390a0..ce6d334 100644
--- a/ui_curses.c
+++ b/ui_curses.c
@@ -104,7 +104,9 @@ static char print_buffer[512];
 /* destination buffer for utf8_encode and utf8_decode */
 static char conv_buffer[512];
 
-#define print_buffer_size (sizeof(print_buffer) - 1)
+/* one character can take up to 4 bytes in UTF-8 */
+#define print_buffer_max_width (sizeof(print_buffer) / 4 - 1)
+
 static int using_utf8;
 
 static const char *t_ts;
@@ -1181,10 +1183,10 @@ static void do_update_titleline(void)
 
 		/* set window title */
 		if (use_alt_format) {
-			format_print(print_buffer, sizeof(print_buffer) - 1,
+			format_print(print_buffer, print_buffer_max_width,
 					window_title_alt_format, track_fopts);
 		} else {
-			format_print(print_buffer, sizeof(print_buffer) - 1,
+			format_print(print_buffer,  print_buffer_max_width,
 					window_title_format, track_fopts);
 		}
 
-- 
1.6.1

